router/pkg/metric/cache_metrics.go

```diff
@@ -0,0 +1,200 @@
+)
+
+// CacheMetricInfo is a struct that aggregates information to provide metrics for a single cache implementation.
+type CacheMetricInfo struct {
+	CacheType string
+	MaxCost   int64
+	Metrics   *ristretto.Metrics
+}
+
+// NewCacheMetricInfo creates a new CacheMetricInfo instance.
+func NewCacheMetricInfo(cacheType string, maxCost int64, cacheMetrics *ristretto.Metrics) CacheMetricInfo {
+	return CacheMetricInfo{
+		CacheType: cacheType,
+		MaxCost:   maxCost,
+		Metrics:   cacheMetrics,
+	}
+}
+
+var (
+	cacheHitStats  otelmetric.Int64ObservableCounter
+	cacheKeyStats  otelmetric.Int64ObservableCounter
+	cacheCostStats otelmetric.Int64ObservableCounter
+	cacheMaxCost   otelmetric.Int64ObservableGauge
+)
+
+// CacheMetrics is a struct that holds the metrics for various graphql operation caches.
+type CacheMetrics struct {
+	meters                  []otelmetric.Meter
+	baseAttributes          []attribute.KeyValue
+	instrumentRegistrations []otelmetric.Registration
+	logger                  *zap.Logger
+}
+
+// NewCacheMetrics creates a new CacheMetrics instance.
+func NewCacheMetrics(logger *zap.Logger, baseAttributes []attribute.KeyValue, providers ...*metric.MeterProvider) (*CacheMetrics, error) {
+	configuredMeters := make([]otelmetric.Meter, 0, len(providers))
+
+	for _, meterProvider := range providers {
+		meter := meterProvider.Meter(cosmoRouterCacheMeterName,
+			otelmetric.WithInstrumentationVersion(cosmoRouterCacheMeterVersion))
+
+		if err := configureMeter(meter); err != nil {
+			return nil, err
+		}
+
+		configuredMeters = append(configuredMeters, meter)
+	}
+
+	return &CacheMetrics{
+		meters:         configuredMeters,
+		baseAttributes: baseAttributes,
+		logger:         logger,
+	}, nil
+}
+
+func configureMeter(meter otelmetric.Meter) error {
+	var err error
+	if cacheHitStats, err = meter.Int64ObservableCounter(
+		operationCacheHitMetric,
+		otelmetric.WithDescription("Cache stats related to cache access. Tracks cache hits and misses. Can be used to calculate the ratio"),
+	); err != nil {
+		return err
+	}
+
+	if cacheKeyStats, err = meter.Int64ObservableCounter(
+		operationCacheKeyMetric,
+		otelmetric.WithDescription("Cache stats for Keys. Tracks added, updated and evicted keys. Can be used to get the total number of items"),
+	); err != nil {
+		return err
+	}
+
+	if cacheCostStats, err = meter.Int64ObservableCounter(
+		operationCacheCostMetric,
+		otelmetric.WithDescription("Cache stats for Cost. Tracks the cost of the cache operations. Can be used to calculate the cost of the cache operations"),
+	); err != nil {
+		return err
+	}
+
+	if cacheMaxCost, err = meter.Int64ObservableGauge(
+		operationCacheCostMaxMetric,
+		otelmetric.WithDescription("Tracks the maximum configured cost for a cache. Useful to investigate differences between the number of keys and the current cost"),
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *CacheMetrics) Observe(
```

Comment:
the func name Observe indicates that this func will be blocking. When looking closer, it appears that this is not the case. Maybe different naming could clarify this?

Alternatively, you could add comments to the functions to make the functionality clear.